scipy/optimize/tests/test_optimize.py

```diff
@@ -1238,6 +1240,15 @@ def test_respect_maxiter(self, method):
         if method == 'SLSQP':
             assert sol.status == 9  # Iteration limit reached
 
+    @pytest.mark.parametrize('method', ['Nelder-Mead', 'Powell'])
+    def test_user_warning(self, method):
+        x0 = np.zeros(10)
+        sf = ScalarFunction(optimize.rosen, x0, (), optimize.rosen_der,
+                            optimize.rosen_hess, None, None)
+        with pytest.warns(UserWarning, match=r'Maximum number of iterations'):
```

Comment:
It seems more appropriate to raise a `RuntimeWarning` for cases like this. What do you think?

Reply:
I'd agree. From [here](https://docs.python.org/3/library/exceptions.html#UserWarning):
UserWarning - Base class for warnings generated by user code.
RuntimeWarning - Base class for warnings about dubious runtime behavior.
The latter seems more appropriate.

Reply:
Fixed.